opt_adapt/opt.py

```diff
@@ -52,11 +53,33 @@ def __init__(self, options={}):
                 raise ValueError(f"Option {key} not recognised")
             self.__setattr__(key, value)
 
+# Only works for problem with control paramter in r-space
+def line_search(forward_run, mesh, u, P, J, dJ, Rspace, alpha=1e-4, max_search_iter=100):
+    
+    lr = 1
+    if Rspace:
+        initial_slope = float(dJ) * float(P)
+    else:
+        initial_slope = np.dot(dJ.dat.data, P.dat.data)
```

Comment:
You don't need the `if Rspace` condition because `np.dot(dJ.dat.data, P.dat.data)` will give the same result in both cases.